File: PolymarketBot/src/polymarket_bot/market_data.py
from __future__ import annotations

import httpx
from py_clob_client.client import ClobClient
from py_clob_client.clob_types import BookParams

GAMMA_API = "https://gamma-api.polymarket.com"


def get_active_markets(limit: int = 10, offset: int = 0) -> list[dict]:
    """Fetch active markets from the Gamma REST API (no auth required)."""
    resp = httpx.get(
        f"{GAMMA_API}/markets",
        params={"active": "true", "closed": "false", "limit": limit, "offset": offset},
        timeout=10,
    )
    resp.raise_for_status()
    return resp.json()


def get_market_detail(clob: ClobClient, condition_id: str) -> dict:
    """Fetch single market detail from CLOB (includes tick size, neg_risk)."""
    return clob.get_market(condition_id)


def get_order_book(clob: ClobClient, token_id: str):
    """Fetch the order book for a given outcome token (returns OrderBookSummary)."""
    return clob.get_order_book(token_id)


def get_order_books(clob: ClobClient, token_ids: list[str]) -> list:
    """Fetch order books for multiple outcome tokens in one call."""
    params = [BookParams(token_id=tid) for tid in token_ids]
    return clob.get_order_books(params)


def get_midpoint(clob: ClobClient, token_id: str) -> dict:
    """Return the mid-market price (average of best bid and ask)."""
    return clob.get_midpoint(token_id)


def get_spread(clob: ClobClient, token_id: str) -> dict:
    """Return the bid-ask spread for a token."""
    return clob.get_spread(token_id)


def get_last_trade_price(clob: ClobClient, token_id: str) -> dict:
    """Return the most recent trade price for a token."""
    return clob.get_last_trade_price(token_id)

File: PolymarketBot/functions/src/marketData.ts
import { ClobClient } from "@polymarket/clob-client";

const GAMMA_API = "https://gamma-api.polymarket.com";

export async function getActiveMarkets(
  limit  = 10,
  offset = 0,
): Promise<Record<string, unknown>[]> {
  const url = new URL(`${GAMMA_API}/markets`);
  url.searchParams.set("active",  "true");
  url.searchParams.set("closed",  "false");
  url.searchParams.set("limit",   String(limit));
  url.searchParams.set("offset",  String(offset));

  const resp = await fetch(url.toString());
  if (!resp.ok) throw new Error(`Gamma API error: ${resp.status} ${resp.statusText}`);
  return resp.json() as Promise<Record<string, unknown>[]>;
}

export async function getOrderBook(clob: ClobClient, tokenId: string) {
  return clob.getOrderBook(tokenId);
}

export async function getMidpoint(
  clob:    ClobClient,
  tokenId: string,
): Promise<{ mid: string }> {
  return clob.getMidpoint(tokenId) as Promise<{ mid: string }>;
}

export async function getSpread(
  clob:    ClobClient,
  tokenId: string,
): Promise<{ spread: string }> {
  return clob.getSpread(tokenId) as Promise<{ spread: string }>;
}
